File: server/user-service/src/main/java/com/mshop/userservice/controller/dto/UpsertUserProductDto.java
package com.mshop.userservice.controller.dto;

import lombok.AllArgsConstructor;
import lombok.Data;

@Data
@AllArgsConstructor
public class UpsertUserProductDto {
    private String productId;
    private String userId;

    public UpsertUserProductDto(String productId) {
        this.productId = productId;
        this.userId = null;
    }
}

File: server/product-service/src/main/kotlin/com/mshop/productservice/dto/UpsertUserProductDto.kt
package com.mshop.productservice.dto

import java.time.LocalDateTime

data class UpsertUserProductDto(
    val userProductId: String? = null,
    val productId: String? = null,
    val userId: String? = null,
    val rate: Float? = null,
    val matchRate: Float? = null,
    val avgRate: Float? = null,
    val comment: String? = null,
    var createdAt: LocalDateTime? = null,
    var images: List<String>? = null,
) {
    companion object {
        fun create(userId: String, productId: String): UpsertUserProductDto {
            return UpsertUserProductDto(
                productId = productId,
                userId = userId
            )
        }
    }
}